File: dsa-problems/Hashing/solutions/javascript/HSH-006-minimal-rotation-hash.js
const readline = require("readline");

class Solution {
  minimalRotation(s) {
    const n = s.length;
    const doubled = s + s;
    const m = doubled.length;
    
    const MOD = 1000000007n;
    const BASE = 313n;
    
    const h = new BigInt64Array(m + 1);
    const p = new BigInt64Array(m + 1);
    p[0] = 1n;
    
    for (let i = 0; i < m; i++) {
      const code = BigInt(doubled.charCodeAt(i));
      h[i + 1] = (h[i] * BASE + code) % MOD;
      p[i + 1] = (p[i] * BASE) % MOD;
    }
    
    const getHash = (l, r) => {
      const len = r - l + 1;
      let val = (h[r + 1] - (h[l] * p[len]) % MOD) % MOD;
      if (val < 0n) val += MOD;
      return val;
    };
    
    const getLCP = (i, j) => {
      let low = 0, high = n;
      let ans = 0;
      while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (mid === 0) {
          low = mid + 1;
          continue;
        }
        
        const h1 = getHash(i, i + mid - 1);
        const h2 = getHash(j, j + mid - 1);
        
        if (h1 === h2) {
          ans = mid;
          low = mid + 1;
        } else {
          high = mid - 1;
        }
      }
      return ans;
    };
    
    let best = 0;
    for (let curr = 1; curr < n; curr++) {
      const lcp = getLCP(best, curr);
      if (lcp < n) {
        if (doubled.charCodeAt(curr + lcp) < doubled.charCodeAt(best + lcp)) {
          best = curr;
        }
      }
    }
    
    return doubled.substring(best, best + n);
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(line.trim()));
rl.on("close", () => {
  if (data.length === 0) return;
  const s = data[0];

  const solution = new Solution();
  console.log(solution.minimalRotation(s));
});

File: dsa-problems/Hashing/solutions/python/HSH-006-minimal-rotation-hash.py
import sys

# Increase recursion depth just in case
sys.setrecursionlimit(2000)

class Solution:
    def minimal_rotation(self, s: str) -> str:
        n = len(s)
        doubled = s + s
        m = len(doubled)
        
        MOD = 10**9 + 7
        BASE = 313
        
        h = [0] * (m + 1)
        p = [1] * (m + 1)
        
        for i in range(m):
            h[i+1] = (h[i] * BASE + ord(doubled[i])) % MOD
            p[i+1] = (p[i] * BASE) % MOD
            
        def get_hash(l, r):
            length = r - l + 1
            return (h[r+1] - h[l] * p[length]) % MOD
            
        def get_lcp(i, j):
            low, high = 0, n
            ans = 0
            while low <= high:
                mid = (low + high) // 2
                if mid == 0:
                    low = mid + 1
                    continue
                
                h1 = get_hash(i, i + mid - 1)
                h2 = get_hash(j, j + mid - 1)
                
                if h1 == h2:
                    ans = mid
                    low = mid + 1
                else:
                    high = mid - 1
            return ans
            
        best = 0
        for curr in range(1, n):
            lcp = get_lcp(best, curr)
            if lcp < n:
                if doubled[curr + lcp] < doubled[best + lcp]:
                    best = curr
                    
        return doubled[best : best + n]

def minimal_rotation(s: str) -> str:
    solver = Solution()
    return solver.minimal_rotation(s)

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    s = input_data[0]
    print(minimal_rotation(s))

if __name__ == "__main__":
    main()
